--- /tmp/isa_export.ml	2017-04-24 16:55:22.354390989 +0100
+++ isa_export.ml.patched	2017-04-24 16:57:55.526490791 +0100
@@ -876,7 +876,7 @@ let rec dest_Node_frame
 
 end;; (*struct Frame*)
 
-module Params : sig
+module type PARAMS = sig
   type k
   val equal_k : k HOL.equal
   type v
@@ -891,7 +891,7 @@ module Params : sig
   val store_free : page_ref list -> unit mm
   val store_read : page_ref -> (k, v, page_ref) Frame.t mm
   val store_alloc : (k, v, page_ref) Frame.t -> page_ref mm
-end = struct
+end (* = struct
 
 type k = K of Arith.nat;;
 
@@ -928,7 +928,9 @@ let rec store_read r = Util.failwitha "F
 
 let rec store_alloc frm = Util.failwitha "FIXME";;
 
-end;; (*struct Params*)
+end;; *) (*struct Params*)
+
+module Make = functor (Params:PARAMS) -> struct
 
 module Monad : sig
   val dest_MM : 'a Params.mm -> Params.store -> Params.store * 'a Util.res
@@ -1974,3 +1976,5 @@ let rec lss_is_finished
           | Pre_params.LS_up (_ :: _) -> false);;
 
 end;; (*struct Leaf_stream*)
+
+end
\ No newline at end of file
