--- isa_export_recap.mli.orig	2018-03-08 12:01:58.635748288 +0000
+++ isa_export_recap.mli.patched	2018-03-08 12:05:16.984139272 +0000
@@ -11,9 +11,7 @@
           Tjr_btree.Isa_export.Arith.nat * Tjr_btree.Isa_export.Arith.nat *
           Tjr_btree.Isa_export.Arith.nat * 'a
     val max_leaf_size : 'a constants_ext -> Tjr_btree.Isa_export.Arith.nat
-    val max_node_keys : 'a constants_ext -> Tjr_btree.Isa_export.Arith.nat
-    val min_leaf_size : 'a constants_ext -> Tjr_btree.Isa_export.Arith.nat
-    val min_node_keys : 'a constants_ext -> Tjr_btree.Isa_export.Arith.nat
+    ...
   end
 module Tree :
   sig
@@ -21,23 +19,7 @@
       ('a, 'b) Tjr_btree.Isa_export.Tree.tree =
         Node of ('a list * ('a, 'b) tree list)
       | Leaf of ('a * 'b) list
-    val tree_to_yojson :
-      ('a -> Yojson.Safe.json) ->
-      ('b -> Yojson.Safe.json) -> ('a, 'b) tree -> Yojson.Safe.json
-    val tree_of_yojson :
-      (Yojson.Safe.json -> 'a Ppx_deriving_yojson_runtime.error_or) ->
-      (Yojson.Safe.json -> 'b Ppx_deriving_yojson_runtime.error_or) ->
-      Yojson.Safe.json -> ('a, 'b) tree Ppx_deriving_yojson_runtime.error_or
-    val height : ('a, 'b) tree -> Tjr_btree.Isa_export.Arith.nat
-    val dest_Node : ('a, 'b) tree -> 'a list * ('a, 'b) tree list
-    val tree_equal : ('a, 'b) tree -> ('a, 'b) tree -> bool
-    val tree_to_leaves : ('a, 'b) tree -> ('a * 'b) list list
-    val tree_to_kvs : ('a, 'b) tree -> ('a * 'b) list
-    val tree_to_keys : ('a, 'b) tree -> 'a Tjr_btree.Isa_export.Set.set
-    val wellformed_tree :
-      unit Tjr_btree.Isa_export.Prelude.constants_ext ->
-      Tjr_btree.Isa_export.Prelude.min_size_t option ->
-      ('a -> 'a -> Tjr_btree.Isa_export.Arith.int) -> ('a, 'b) tree -> bool
+    ...
   end
 module Tree_stack :
   sig
@@ -92,19 +74,7 @@
       ('a, 'b, 'c) Tjr_btree.Isa_export.Disk_node.dnode =
         Disk_node of ('a list * 'c list)
       | Disk_leaf of ('a * 'b) list
-    val dnode_to_yojson :
-      ('a -> Yojson.Safe.json) ->
-      ('b -> Yojson.Safe.json) ->
-      ('c -> Yojson.Safe.json) -> ('a, 'b, 'c) dnode -> Yojson.Safe.json
-    val dnode_of_yojson :
-      (Yojson.Safe.json -> 'a Ppx_deriving_yojson_runtime.error_or) ->
-      (Yojson.Safe.json -> 'b Ppx_deriving_yojson_runtime.error_or) ->
-      (Yojson.Safe.json -> 'c Ppx_deriving_yojson_runtime.error_or) ->
-      Yojson.Safe.json ->
-      ('a, 'b, 'c) dnode Ppx_deriving_yojson_runtime.error_or
-    val mk_Disk_node : 'a list * 'b list -> ('a, 'c, 'b) dnode
-    val dest_Disk_leaf : ('a, 'b, 'c) dnode -> ('a * 'b) list
-    val dest_Disk_node : ('a, 'b, 'c) dnode -> 'a list * 'c list
+    ...
   end
 module Params :
   sig
@@ -151,38 +121,11 @@
   sig
     type ('a, 'b, 'c) find_state =
         ('a, 'b, 'c) Tjr_btree.Isa_export.Find.find_state
-    val find_state_to_yojson :
-      ('a -> Yojson.Safe.json) ->
-      ('b -> Yojson.Safe.json) ->
-      ('c -> Yojson.Safe.json) -> ('a, 'b, 'c) find_state -> Yojson.Safe.json
-    val find_state_of_yojson :
-      (Yojson.Safe.json -> 'a Ppx_deriving_yojson_runtime.error_or) ->
-      (Yojson.Safe.json -> 'b Ppx_deriving_yojson_runtime.error_or) ->
-      (Yojson.Safe.json -> 'c Ppx_deriving_yojson_runtime.error_or) ->
-      Yojson.Safe.json ->
-      ('a, 'b, 'c) find_state Ppx_deriving_yojson_runtime.error_or
     val find_step :
       ('a, 'b, 'c, 'd) Tjr_btree.Isa_export.Params.ps1 ->
       ('a, 'b, 'c) find_state ->
       'd -> 'd * ('a, 'b, 'c) find_state Tjr_btree.Isa_export.Util.res
-    val mk_find_state : 'a -> 'b -> ('a, 'c, 'b) find_state
-    val wf_store_tree :
-      ('a -> 'b -> ('c, 'd) Tjr_btree.Isa_export.Tree.tree option) ->
-      'a -> 'b -> ('c, 'd) Tjr_btree.Isa_export.Tree.tree -> bool
-    val dest_f_finished :
-      ('a, 'b, 'c) find_state ->
-      ('c *
-       ('a *
-        ('c *
-         (('a * 'b) list *
-          ('a, 'c, unit)
-          Tjr_btree.Isa_export.Searching_and_splitting.rsplit_node_ext list))))
-      option
-    val wellformed_find_state :
-      ('a -> 'a -> Tjr_btree.Isa_export.Arith.int) ->
-      ('b -> 'c -> ('a, 'd) Tjr_btree.Isa_export.Tree.tree option) ->
-      ('a, 'd) Tjr_btree.Isa_export.Tree.tree ->
-      'b -> ('a, 'd, 'c) find_state -> bool
+    ...
   end
 module Pre_insert :
   sig
@@ -199,16 +142,7 @@
       ('a, 'b, 'c) Tjr_btree.Isa_export.Insert.i12_t =
         I1 of 'c
       | I2 of ('c * ('a * 'c))
-    val i12_t_to_yojson :
-      ('a -> Yojson.Safe.json) ->
-      ('b -> Yojson.Safe.json) ->
-      ('c -> Yojson.Safe.json) -> ('a, 'b, 'c) i12_t -> Yojson.Safe.json
-    val i12_t_of_yojson :
-      (Yojson.Safe.json -> 'a Ppx_deriving_yojson_runtime.error_or) ->
-      (Yojson.Safe.json -> 'b Ppx_deriving_yojson_runtime.error_or) ->
-      (Yojson.Safe.json -> 'c Ppx_deriving_yojson_runtime.error_or) ->
-      Yojson.Safe.json ->
-      ('a, 'b, 'c) i12_t Ppx_deriving_yojson_runtime.error_or
+...
     type ('a, 'b, 'c) insert_state =
       ('a, 'b, 'c) Tjr_btree.Isa_export.Insert.insert_state =
         I_down of (('a, 'b, 'c) Tjr_btree.Isa_export.Find.find_state * 'b)
@@ -217,29 +151,11 @@
            ('a, 'c, unit)
            Tjr_btree.Isa_export.Searching_and_splitting.rsplit_node_ext list)
       | I_finished of 'c
-    val insert_state_to_yojson :
-      ('a -> Yojson.Safe.json) ->
-      ('b -> Yojson.Safe.json) ->
-      ('c -> Yojson.Safe.json) ->
-      ('a, 'b, 'c) insert_state -> Yojson.Safe.json
-    val insert_state_of_yojson :
-      (Yojson.Safe.json -> 'a Ppx_deriving_yojson_runtime.error_or) ->
-      (Yojson.Safe.json -> 'b Ppx_deriving_yojson_runtime.error_or) ->
-      (Yojson.Safe.json -> 'c Ppx_deriving_yojson_runtime.error_or) ->
-      Yojson.Safe.json ->
-      ('a, 'b, 'c) insert_state Ppx_deriving_yojson_runtime.error_or
     val insert_step :
       ('a, 'b, 'c, 'd) Tjr_btree.Isa_export.Params.ps1 ->
       ('a, 'b, 'c) insert_state ->
       'd -> 'd * ('a, 'b, 'c) insert_state Tjr_btree.Isa_export.Util.res
-    val dest_i_finished : ('a, 'b, 'c) insert_state -> 'c option
-    val mk_insert_state : 'a -> 'b -> 'c -> ('a, 'b, 'c) insert_state
-    val wellformed_insert_state :
-      unit Tjr_btree.Isa_export.Prelude.constants_ext ->
-      ('a -> 'a -> Tjr_btree.Isa_export.Arith.int) ->
-      ('b -> 'c -> ('a, 'd) Tjr_btree.Isa_export.Tree.tree option) ->
-      ('a, 'd) Tjr_btree.Isa_export.Tree.tree ->
-      'b -> 'a -> 'd -> ('a, 'd, 'c) insert_state -> bool
+...
   end
 module Delete2 :
   sig
@@ -248,16 +164,6 @@
         D_small_leaf of ('a * 'b) list
       | D_small_node of ('a list * 'c list)
       | D_updated_subtree of 'c
-    val del_t_to_yojson :
-      ('a -> Yojson.Safe.json) ->
-      ('b -> Yojson.Safe.json) ->
-      ('c -> Yojson.Safe.json) -> ('a, 'b, 'c) del_t -> Yojson.Safe.json
-    val del_t_of_yojson :
-      (Yojson.Safe.json -> 'a Ppx_deriving_yojson_runtime.error_or) ->
-      (Yojson.Safe.json -> 'b Ppx_deriving_yojson_runtime.error_or) ->
-      (Yojson.Safe.json -> 'c Ppx_deriving_yojson_runtime.error_or) ->
-      Yojson.Safe.json ->
-      ('a, 'b, 'c) del_t Ppx_deriving_yojson_runtime.error_or
     type ('a, 'b, 'c) delete_state =
       ('a, 'b, 'c) Tjr_btree.Isa_export.Delete2.delete_state =
         D_down of (('a, 'b, 'c) Tjr_btree.Isa_export.Find.find_state * 'c)
@@ -267,29 +173,12 @@
             Tjr_btree.Isa_export.Searching_and_splitting.rsplit_node_ext list *
             'c))
       | D_finished of 'c
-    val delete_state_to_yojson :
-      ('a -> Yojson.Safe.json) ->
-      ('b -> Yojson.Safe.json) ->
-      ('c -> Yojson.Safe.json) ->
-      ('a, 'b, 'c) delete_state -> Yojson.Safe.json
-    val delete_state_of_yojson :
-      (Yojson.Safe.json -> 'a Ppx_deriving_yojson_runtime.error_or) ->
-      (Yojson.Safe.json -> 'b Ppx_deriving_yojson_runtime.error_or) ->
-      (Yojson.Safe.json -> 'c Ppx_deriving_yojson_runtime.error_or) ->
-      Yojson.Safe.json ->
-      ('a, 'b, 'c) delete_state Ppx_deriving_yojson_runtime.error_or
+...
     val delete_step :
       ('a, 'b, 'c, 'd) Tjr_btree.Isa_export.Params.ps1 ->
       ('a, 'b, 'c) delete_state ->
       'd -> 'd * ('a, 'b, 'c) delete_state Tjr_btree.Isa_export.Util.res
-    val dest_d_finished : ('a, 'b, 'c) delete_state -> 'c option
-    val mk_delete_state : 'a -> 'b -> ('a, 'c, 'b) delete_state
-    val wellformed_delete_state :
-      unit Tjr_btree.Isa_export.Prelude.constants_ext ->
-      ('a -> 'a -> Tjr_btree.Isa_export.Arith.int) ->
-      ('b -> 'c -> ('a, 'd) Tjr_btree.Isa_export.Tree.tree option) ->
-      ('a, 'd) Tjr_btree.Isa_export.Tree.tree ->
-      'b -> 'a -> ('a, 'd, 'c) delete_state -> bool
+...
   end
 module Insert_many :
   sig
@@ -311,9 +200,7 @@
       ('a, 'b, 'c, 'd) Tjr_btree.Isa_export.Params.ps1 ->
       ('a, 'b, 'c) ist ->
       'd -> 'd * ('a, 'b, 'c) ist Tjr_btree.Isa_export.Util.res
-    val dest_i_finished : ('a, 'b, 'c) ist -> ('c * ('a * 'b) list) option
-    val mk_insert_state :
-      'a -> 'b -> ('a * 'b) list -> 'c -> ('a, 'b, 'c) ist
+...
   end
 module Leaf_stream :
   sig
@@ -323,7 +210,5 @@
       ('a, 'b, 'c, 'd) Tjr_btree.Isa_export.Params.ps1 ->
       ('a, 'b, 'c) ls_state ->
       'd -> 'd * ('a, 'b, 'c) ls_state Tjr_btree.Isa_export.Util.res
-    val mk_ls_state : 'a -> ('b, 'c, 'a) ls_state
-    val dest_LS_leaf : ('a, 'b, 'c) ls_state -> ('a * 'b) list option
-    val lss_is_finished : ('a, 'b, 'c) ls_state -> bool
+...
   end
