--- orig.ml	2017-12-13 14:05:35.198604441 +0000
+++ patched.ml	2017-12-13 14:20:35.992354711 +0000
@@ -1,3 +1,7 @@
+(** This file is exported from Isabelle, and lightly patched (eg to
+   include this comment!). The OCaml interfaces wrap this basic
+   functionality. *)
+
 module Fun : sig
   val id : 'a -> 'a
   val comp : ('a -> 'b) -> ('c -> 'a) -> 'c -> 'b
@@ -271,9 +275,16 @@ let rec is_none = function Some x -> fal
 
 end;; (*struct Option*)
 
+module Res = struct
+  include Pervasives
+  type 'a res = ('a,string) result
+end
+
+let check_flag = ref true
+
 module Util : sig
+  include module type of Res
   type error = String_error of string
-  type 'a res = Ok of 'a | Error of error
   val from_to : Arith.nat -> Arith.nat -> Arith.nat list
   val is_Nil : 'a list -> bool
   val is_None : 'a option -> bool
@@ -294,7 +305,7 @@ end = struct
 
 type error = String_error of string;;
 
-type 'a res = Ok of 'a | Error of error;;
+include Res
 
 let rec from_to x y = List.upt x (Arith.suc y);;
 
@@ -304,9 +315,9 @@ let rec is_None x = Option.is_none x;;
 
 let rec rev_apply x f = f x;;
 
-let rec failwitha x = rev_apply "FIXME patch" (fun _ -> failwith "undefined");;
+let rec failwitha x = failwith x
 
-let rec check_true f = rev_apply "FIXME patch" (fun _ -> failwith "undefined");;
+let rec check_true f = if !check_flag then f () else true;;
 
 let rec split_at
   n xs =
@@ -498,7 +509,7 @@ let ck2_tests : unit
       in
     ();;
 
-let rec kvs_equal x = Util.failwitha "FIXME patch" x;;
+let rec kvs_equal x y = (x=y)
 
 let rec ordered_key_list
   ord ks =
@@ -582,7 +593,7 @@ end;; (*struct Key_value*)
 
 module Searching_and_splitting : sig
   type ('a, 'b, 'c) rsplit_node_ext =
-    Rsplit_node_ext of 'a list * 'b list * 'b * 'a list * 'b list * 'c
+    Rsplit_node_ext of 'a list * 'b list * 'b * 'a list * 'b list * 'c  [@@deriving yojson]
   val r_ts2 : ('a, 'b, 'c) rsplit_node_ext -> 'b list
   val r_ts1 : ('a, 'b, 'c) rsplit_node_ext -> 'b list
   val r_ks2 : ('a, 'b, 'c) rsplit_node_ext -> 'a list
@@ -617,7 +628,7 @@ module Searching_and_splitting : sig
 end = struct
 
 type ('a, 'b, 'c) rsplit_node_ext =
-  Rsplit_node_ext of 'a list * 'b list * 'b * 'a list * 'b list * 'c;;
+  Rsplit_node_ext of 'a list * 'b list * 'b * 'a list * 'b list * 'c  [@@deriving yojson];;
 
 let rec aux
   cmp k0 ks_rs1 ks_rs2 =
@@ -740,7 +751,7 @@ end;; (*struct Prelude*)
 
 module Tree : sig
   type ('a, 'b) tree = Node of ('a list * ('a, 'b) tree list) |
-    Leaf of ('a * 'b) list
+    Leaf of ('a * 'b) list [@@deriving yojson]
   val height : ('a, 'b) tree -> Arith.nat
   val dest_Node : ('a, 'b) tree -> 'a list * ('a, 'b) tree list
   val tree_equal : ('a, 'b) tree -> ('a, 'b) tree -> bool
@@ -754,7 +765,7 @@ module Tree : sig
 end = struct
 
 type ('a, 'b) tree = Node of ('a list * ('a, 'b) tree list) |
-  Leaf of ('a * 'b) list;;
+  Leaf of ('a * 'b) list  [@@deriving yojson];;
 
 let rec tree_to_subtrees
   t0 = (match t0
@@ -849,7 +860,7 @@ let rec dest_Node
   = function Node (ks, rs) -> (ks, rs)
     | Leaf uu -> Util.failwitha "dest_Node";;
 
-let rec tree_equal t1 t2 = Util.failwitha "FIXME patch";;
+let rec tree_equal t1 t2 = (t1=t2)
 
 let rec tree_to_leaves
   t0 = (match t0
@@ -966,7 +977,7 @@ let rec no_focus
 let rec r_stk_to_rs
   xs = Util.rev_apply xs (List.map Searching_and_splitting.r_t);;
 
-let rec rstack_equal s1 s2 = Util.failwitha "FIXME patch";;
+let rec rstack_equal s1 s2 = (s1=s2)
 
 let rec tree_to_rstack
   ord k t n =
@@ -1007,14 +1018,14 @@ end;; (*struct Tree_stack*)
 
 module Disk_node : sig
   type ('a, 'b, 'c) dnode = Disk_node of ('a list * 'c list) |
-    Disk_leaf of ('a * 'b) list
+    Disk_leaf of ('a * 'b) list  [@@deriving yojson]
   val mk_Disk_node : 'a list * 'b list -> ('a, 'c, 'b) dnode
   val dest_Disk_leaf : ('a, 'b, 'c) dnode -> ('a * 'b) list
   val dest_Disk_node : ('a, 'b, 'c) dnode -> 'a list * 'c list
 end = struct
 
 type ('a, 'b, 'c) dnode = Disk_node of ('a list * 'c list) |
-  Disk_leaf of ('a * 'b) list;;
+  Disk_leaf of ('a * 'b) list   [@@deriving yojson];;
 
 let rec check_length_ks_rs
   ks_rs =
@@ -1153,7 +1164,7 @@ let rec return x = (fun s -> (s, Util.Ok
 end;; (*struct Monad*)
 
 module Find : sig
-  type ('a, 'b, 'c) find_state
+  type ('a, 'b, 'c) find_state [@@deriving yojson]
   val find_step :
     ('a, 'b, 'c, 'd) Params.ps1 ->
       ('a, 'b, 'c) find_state -> 'd -> 'd * ('a, 'b, 'c) find_state Util.res
@@ -1179,7 +1190,7 @@ type ('a, 'b, 'c) find_state =
   | F_finished of
       ('c * ('a * ('c * (('a * 'b) list *
                           ('a, 'c, unit)
-                            Searching_and_splitting.rsplit_node_ext list))));;
+                            Searching_and_splitting.rsplit_node_ext list))))  [@@deriving yojson];;
 
 let rec find_step
   ps1 fs =
@@ -1300,12 +1311,12 @@ let rec split_node
 end;; (*struct Pre_insert*)
 
 module Insert : sig
-  type ('a, 'b, 'c) i12_t = I1 of 'c | I2 of ('c * ('a * 'c))
+  type ('a, 'b, 'c) i12_t = I1 of 'c | I2 of ('c * ('a * 'c))  [@@deriving yojson]
   type ('a, 'b, 'c) insert_state = I_down of (('a, 'b, 'c) Find.find_state * 'b)
     | I_up of
         (('a, 'b, 'c) i12_t *
           ('a, 'c, unit) Searching_and_splitting.rsplit_node_ext list)
-    | I_finished of 'c
+    | I_finished of 'c  [@@deriving yojson]
   val insert_step :
     ('a, 'b, 'c, 'd) Params.ps1 ->
       ('a, 'b, 'c) insert_state -> 'd -> 'd * ('a, 'b, 'c) insert_state Util.res
@@ -1319,13 +1330,13 @@ module Insert : sig
             'b -> 'a -> 'd -> ('a, 'd, 'c) insert_state -> bool
 end = struct
 
-type ('a, 'b, 'c) i12_t = I1 of 'c | I2 of ('c * ('a * 'c));;
+type ('a, 'b, 'c) i12_t = I1 of 'c | I2 of ('c * ('a * 'c))  [@@deriving yojson];;
 
 type ('a, 'b, 'c) insert_state = I_down of (('a, 'b, 'c) Find.find_state * 'b) |
   I_up of
     (('a, 'b, 'c) i12_t *
       ('a, 'c, unit) Searching_and_splitting.rsplit_node_ext list)
-  | I_finished of 'c;;
+  | I_finished of 'c  [@@deriving yojson];;
 
 let rec wf_d
   k_ord r2t t0 s d =
@@ -1525,12 +1536,12 @@ end;; (*struct Insert*)
 
 module Delete2 : sig
   type ('a, 'b, 'c) del_t = D_small_leaf of ('a * 'b) list |
-    D_small_node of ('a list * 'c list) | D_updated_subtree of 'c
+    D_small_node of ('a list * 'c list) | D_updated_subtree of 'c  [@@deriving yojson]
   type ('a, 'b, 'c) delete_state = D_down of (('a, 'b, 'c) Find.find_state * 'c)
     | D_up of
         (('a, 'b, 'c) del_t *
           (('a, 'c, unit) Searching_and_splitting.rsplit_node_ext list * 'c))
-    | D_finished of 'c
+    | D_finished of 'c  [@@deriving yojson]
   val delete_step :
     ('a, 'b, 'c, 'd) Params.ps1 ->
       ('a, 'b, 'c) delete_state -> 'd -> 'd * ('a, 'b, 'c) delete_state Util.res
@@ -1544,13 +1555,13 @@ module Delete2 : sig
 end = struct
 
 type ('a, 'b, 'c) del_t = D_small_leaf of ('a * 'b) list |
-  D_small_node of ('a list * 'c list) | D_updated_subtree of 'c;;
+  D_small_node of ('a list * 'c list) | D_updated_subtree of 'c  [@@deriving yojson];;
 
 type ('a, 'b, 'c) delete_state = D_down of (('a, 'b, 'c) Find.find_state * 'c) |
   D_up of
     (('a, 'b, 'c) del_t *
       (('a, 'c, unit) Searching_and_splitting.rsplit_node_ext list * 'c))
-  | D_finished of 'c;;
+  | D_finished of 'c  [@@deriving yojson];;
 
 let rec wf_d
   k_ord r2f t0 s d =
